core/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -152,7 +152,7 @@ public void search(Query query, Collector collector) throws IOException {
             // throws Lucene.EarlyTerminationException when given count is reached
             collector = Lucene.wrapCountBasedEarlyTerminatingCollector(collector, searchContext.terminateAfter());
         }
-        if (currentState == Stage.MAIN_QUERY) {
+        if (currentState == Stage.MAIN_QUERY && (query == searchContext.query() || query == searchContext.parsedQuery().query())) {
```

Comment:
agreed, it is the only scary bit of the PR. I'll try it out.

Without it the old paren/child impl causes stack overflow errors. Running `mvn clean verify` with the PR does execute successfully.